In [ ]:
import numpy as np
import pandas as pd
from twarc import Twarc

In [ ]:
API_KEY = 'DwzCGrRCeMsxoChGP3SbwLvvc'
API_KEY_SECRET = 'iSg172652dagp3Lz846yG5nu8nnB5TJofj0fu94ZRZ68NjLLmR'
ACCESS_TOKEN = '1448836767839367168-34GJVrkILAiqOHhzisyQsvYoIce30J'
ACCESS_TOKEN_SECRET = 'y073DDgfYhUMnseOKf9b7zJkw3Sz0oZtKESSDTYvnItab'

t = Twarc(API_KEY, API_KEY_SECRET, ACCESS_TOKEN, ACCESS_TOKEN_SECRET)


### Dataset 1 - Harvard Dataverse - GWU Libraries

In [ ]:
ids = pd.read_csv('data/Harvard_Dataverse_GW/coronavirus-through-03-December-2020-35.txt')

This dataset contains all COVID-19 related tweets from March through December in 2020. 
March was when shutdowns began in the United States. 

https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/LW0BTB

In [ ]:
ids.head(3)

1330273264137015297
0  1330273265634439171
1  1330273266653650945
2  1330273266729152516

The first id is shown as index. We need to fix that.

In [ ]:
ids = pd.read_csv('/Users/samuelpark/Desktop/Job Search/Bootcamps/BrainStation/Capstone/covid19_tweets/data/Harvard_Dataverse_GW/coronavirus-through-03-December-2020-35.txt', header=None)
#ids.rename(columns={0: 'tweet_id'})
ids.head()

0
0  1330273264137015297
1  1330273265634439171
2  1330273266653650945
3  1330273266729152516
4  1330273266754269186

In [ ]:
ids.shape

(4903485, 1)

The dataset has 4,903,485  rows and 1 column (i.e., it contains 4,903,485 lines of tweet IDs).

In [ ]:
ids[0].unique().size

4901731

Of the almost 5 million rows of tweet IDs, 4,901,731 of them are unique for some reason. 

Note that these are tweet IDs, not user IDs. A unique tweet ID is assigned to a tweet when it is posted online. 



Original `coronavirus-through-03-December-2020-35.txt` has 4903485 rows of tweet IDs.

#### Let's just take out several of the tweet IDs and see what they return upon hydrating.

In [ ]:
test_tweets = pd.read_json('data/harvard_Dataverse_GW/test_tweets.jsonl', lines=True)

In [ ]:
test_tweets

created_at                   id               id_str  \
0 2020-11-21 22:13:56+00:00  1330273264137015297  1330273264137015296   
1 2020-11-21 22:13:56+00:00  1330273266955657223  1330273266955657216   
2 2020-11-21 22:13:57+00:00  1330273269652606978  1330273269652606976   
3 2020-11-21 22:13:56+00:00  1330273265634439171  1330273265634439168   
4 2020-11-21 22:13:56+00:00  1330273266653650945  1330273266653650944   
5 2020-11-21 22:13:56+00:00  1330273266754269186  1330273266754269184   
6 2020-11-21 22:13:57+00:00  1330273267890974720  1330273267890974720   
7 2020-11-21 22:13:57+00:00  1330273268184518660  1330273268184518656   

                                           full_text  truncated  \
0  RT @Mediavenir: 🇫🇷 FLASH - Les #cultes devraie...      False   
1  RT @kerpen: Really encouraging results!  But h...      False   
2  RT @CongressmanRaja: At least 250,000 people i...      False   
3    RT @B52Malmet: This is what Trump has  wrought.      False   
4        #FranceEtatPolicier https://t.co/K3KBJD15g7      False   
5  RT @AiltonBenedito: Escolas de samba do Rio ce...      False   
6  #lockdown #Tyranny #Germany #WhosNext #COVID19...      False   
7  RT @donwinslow: 195,542 NEW #COVID19 cases tod...      False   

  display_text_range                                           entities  \
0           [0, 140]  {'hashtags': [{'text': 'cultes', 'indices': [3...   
1           [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2           [0, 140]  {'hashtags': [{'text': 'COVID19', 'indices': [...   
3            [0, 47]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4            [0, 19]  {'hashtags': [{'text': 'FranceEtatPolicier', '...   
5           [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   
6            [0, 46]  {'hashtags': [{'text': 'lockdown', 'indices': ...   
7           [0, 138]  {'hashtags': [{'text': 'COVID19', 'indices': [...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/iphone" r...                    NaN   
1  <a href="https://mobile.twitter.com" rel="nofo...                    NaN   
2  <a href="http://twitter.com/download/iphone" r...                    NaN   
3  <a href="http://twitter.com/download/android" ...                    NaN   
4  <a href="http://twitter.com/download/android" ...                    NaN   
5  <a href="http://twitter.com/download/android" ...                    NaN   
6  <a href="https://mobile.twitter.com" rel="nofo...                    NaN   
7  <a href="https://mobile.twitter.com" rel="nofo...                    NaN   

   in_reply_to_status_id_str  ...  retweet_count  favorite_count  favorited  \
0                        NaN  ...            128               0      False   
1                        NaN  ...             50               0      False   
2                        NaN  ...            104               0      False   
3                        NaN  ...            235               0      False   
4                        NaN  ...              0               1      False   
5                        NaN  ...           2240               0      False   
6                        NaN  ...              0               0      False   
7                        NaN  ...           2151               0      False   

  retweeted  lang  quoted_status_id  quoted_status_id_str  \
0     False    fr               NaN                   NaN   
1     False    en      1.327333e+18          1.327333e+18   
2     False    en               NaN                   NaN   
3     False    en      1.330267e+18          1.330267e+18   
4     False   und      1.330245e+18          1.330245e+18   
5     False    pt               NaN                   NaN   
6     False   und      1.330195e+18          1.330195e+18   
7     False    en               NaN                   NaN   

                             quoted_status_permalink  \
0                                                

In [ ]:
test_tweets.loc[6, 'full_text']

'#lockdown #Tyranny #Germany #WhosNext #COVID19 https://t.co/f4cYJ7hFtA'

In [ ]:
test_tweets.head().T

0  \
created_at                                         2020-11-21 22:13:56+00:00   
id                                                       1330273264137015297   
id_str                                                   1330273264137015296   
full_text                  RT @Mediavenir: 🇫🇷 FLASH - Les #cultes devraie...   
truncated                                                              False   
display_text_range                                                  [0, 140]   
entities                   {'hashtags': [{'text': 'cultes', 'indices': [3...   
source                     <a href="http://twitter.com/download/iphone" r...   
in_reply_to_status_id                                                    NaN   
in_reply_to_status_id_str                                                NaN   
in_reply_to_user_id                                                      NaN   
in_reply_to_user_id_str                                                  NaN   
in_reply_to_screen_name                                                  NaN   
user                       {'id': 1190319799026487297, 'id_str': '1190319...   
geo                                                                      NaN   
coordinates                                                              NaN   
place                                                                    NaN   
contributors                                                             NaN   
retweeted_status           {'created_at': 'Sat Nov 21 19:52:00 +0000 2020...   
is_quote_status                                                        False   
retweet_count                                                            128   
favorite_count                                                             0   
favorited                                                              False   
retweeted                                                              False   
lang                                                                      fr   
quoted_status_id                                                         NaN   
quoted_status_id_str                                                     NaN   
quoted_status_permalink                                                  NaN   
quoted_status                                                            NaN   
possibly_sensitive                                                       NaN   

                                                                           1  \
created_at                                         2020-11-21 22:13:56+00:00   
id                                                       1330273266955657223   
id_str                                                   1330273266955657216   
full_text                  RT @kerpen: Really encouraging results!  But h...   
truncated                                                              False   
display_text_range                                                  [0, 140]   
entities                   {'hashtags': [], 'symbols': [], 'user_mentions...   
source                     <a href="https://mobile.twitter.com" rel="nofo...   
in_reply_to_status_id                                                    NaN   
in_reply_to_status_id_str                                                NaN   
in_reply_to_user_id                                                      NaN   
in_reply_to_user_id_str                                                  NaN   
in_reply_to_screen_name                                                  NaN   
user                       {'id': 531565758, 'id_str': '531565758', 'name...   
geo                                                                      NaN   
coordinates                                                              NaN   
place                                                                    NaN   
contributors                                                             NaN   
retweeted_status           {'created_at': 'Mon Nov 16 02:47:09 +0000 2020.

In [ ]:
test_tweets.shape

(8, 30)

In [ ]:
test_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 30 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   created_at                 8 non-null      datetime64[ns, UTC]
 1   id                         8 non-null      int64              
 2   id_str                     8 non-null      int64              
 3   full_text                  8 non-null      object             
 4   truncated                  8 non-null      bool               
 5   display_text_range         8 non-null      object             
 6   entities                   8 non-null      object             
 7   source                     8 non-null      object             
 8   in_reply_to_status_id      0 non-null      float64            
 9   in_reply_to_status_id_str  0 non-null      float64            
 10  in_reply_to_user_id        0 non-null      float64            
 11  in_reply_t

---

## Dataset 3 - Gabriel Preda
https://www.kaggle.com/gpreda/covid19-tweets?select=covid19_tweets.csv

In [ ]:
gp = pd.read_csv('data/gabriel_preda/covid19_tweets.csv')
gp.head()

user_name         user_location  \
0           ᏉᎥ☻լꂅϮ            astroworld   
1    Tom Basile 🇺🇸          New York, NY   
2  Time4fisticuffs      Pewee Valley, KY   
3      ethel mertz  Stuck in the Middle    
4         DIPR-J&K     Jammu and Kashmir   

                                    user_description         user_created  \
0  wednesday addams as a disney princess keepin i...  2017-05-26 05:46:42   
1  Husband, Father, Columnist & Commentator. Auth...  2009-04-16 20:06:23   
2  #Christian #Catholic #Conservative #Reagan #Re...  2009-02-28 18:57:41   
3  #Browns #Indians #ClevelandProud #[]_[] #Cavs ...  2019-03-07 01:45:06   
4  🖊️Official Twitter handle of Department of Inf...  2017-02-12 06:45:15   

   user_followers  user_friends  user_favourites  user_verified  \
0             624           950            18775          False   
1            2253          1677               24           True   
2            9275          9525             7254          False   
3             197           987             1488          False   
4          101009           168              101          False   

                  date                                               text  \
0  2020-07-25 12:27:21  If I smelled the scent of hand sanitizers toda...   
1  2020-07-25 12:27:17  Hey @Yankees @YankeesPR and @MLB - wouldn't it...   
2  2020-07-25 12:27:14  @diane3443 @wdunlap @realDonaldTrump Trump nev...   
3  2020-07-25 12:27:10  @brookbanktv The one gift #COVID19 has give me...   
4  2020-07-25 12:27:08  25 July : Media Bulletin on Novel #CoronaVirus...   

                            hashtags               source  is_retweet  
0                                NaN   Twitter for iPhone       False  
1                                NaN  Twitter for Android       False  
2                        ['COVID19']  Twitter for Android       False  
3                        ['COVID19']   Twitter for iPhone       False  
4  ['CoronaVirusUpdates', 'COVID19']  Twitter for Android       False

In [ ]:
gp.loc[2, 'text']

'@diane3443 @wdunlap @realDonaldTrump Trump never once claimed #COVID19 was a hoax. We all claim that this effort to… https://t.co/Jkk8vHWHb3'

---

## Dataset 4 - Shane Smith

In [ ]:
from os import listdir

In [ ]:
path = 'data/shane_smith/2020-april'
# all_files = glob.glob(path +"/*.csv")
# print(all_files)


all_files = [file for file in listdir(path)]
all_files = sorted(all_files) # sort by date

# store 
df_list = []

# tweets = pd.DataFrame()
for filename in all_files:
    # just to show progress
    print(f'reading in {filename}')
    df_temp = pd.read_csv(f'data/shane_smith/2020-april/{filename}', index_col=None, header=0)
    # df = pd.concat([df, df_temp], axis=0, ignore_index=True)
    df_list.append(df_temp)
    print(f'appended dataframe of {filename} to df_list')
    
tweets = pd.concat(df_list, axis=0, ignore_index=True)

reading in 2020-03-29 Coronavirus Tweets.CSV
appended dataframe of 2020-03-29 Coronavirus Tweets.CSV to df_list
reading in 2020-03-30 Coronavirus Tweets.CSV
appended dataframe of 2020-03-30 Coronavirus Tweets.CSV to df_list
reading in 2020-03-31 Coronavirus Tweets.CSV
appended dataframe of 2020-03-31 Coronavirus Tweets.CSV to df_list
reading in 2020-04-01 Coronavirus Tweets.CSV
appended dataframe of 2020-04-01 Coronavirus Tweets.CSV to df_list
reading in 2020-04-02 Coronavirus Tweets.CSV
appended dataframe of 2020-04-02 Coronavirus Tweets.CSV to df_list
reading in 2020-04-03 Coronavirus Tweets.CSV
appended dataframe of 2020-04-03 Coronavirus Tweets.CSV to df_list
reading in 2020-04-04 Coronavirus Tweets.CSV
appended dataframe of 2020-04-04 Coronavirus Tweets.CSV to df_list
reading in 2020-04-05 Coronavirus Tweets.CSV
appended dataframe of 2020-04-05 Coronavirus Tweets.CSV to df_list
reading in 2020-04-06 Coronavirus Tweets.CSV
appended dataframe of 2020-04-06 Coronavirus Tweets.CSV to 

In [ ]:
tweets.shape

(14607013, 22)

In [ ]:
# export the compiled df as csv file
tweets.to_csv(path_or_buf='data/shane_smith/april_tweets_compiled.csv')

In [ ]:
tweets.head()


status_id              user_id            created_at  \
0  1244051646071611394   860252856829587457  2020-03-29T00:00:00Z   
1  1244051645039706112  1125933654943895553  2020-03-29T00:00:00Z   
2  1244051645975191557             80943559  2020-03-29T00:00:00Z   
3  1244051646750928897   817072420947247104  2020-03-29T00:00:00Z   
4  1244051647032102914   788863557349670913  2020-03-29T00:00:00Z   

    screen_name                                               text     source  \
0  IMSS_SanLuis  Ante cualquier enfermedad respiratoria, no te ...  TweetDeck   
1    intrac_ccs  #ATENCIÓN En el Terminal Nuevo Circo se implem...  TweetDeck   
2      rlieving  “People are just storing up. They are staying ...  TweetDeck   
3  Tu_IMSS_Coah  Si empezaste a trabajar, necesitas dar de alta...  TweetDeck   
4  Tabasco_IMSS  Una sociedad informada está mejor preparada an...  TweetDeck   

   reply_to_status_id  reply_to_user_id reply_to_screen_name  is_quote  ...  \
0                 NaN               NaN                  NaN     False  ...   
1                 NaN               NaN                  NaN     False  ...   
2                 NaN               NaN                  NaN     False  ...   
3                 NaN               NaN                  NaN     False  ...   
4                 NaN               NaN                  NaN     False  ...   

   retweet_count  country_code  place_full_name place_type followers_count  \
0              0           NaN              NaN        NaN            1008   
1              1           NaN              NaN        NaN              90   
2              0           NaN              NaN        NaN             136   
3              0           NaN              NaN        NaN            1549   
4              0           NaN              NaN        NaN             868   

  friends_count  account_lang    account_created_at  verified lang  
0            41           NaN  2017-05-04T22:00:38Z     False   es  
1           316           NaN  2019-05-08T01:21:16Z     False   es  
2           457           NaN  2009-10-08T21:06:08Z     False   en  
3           170           NaN  2017-01-05T18:17:00Z     False   es  
4           125           NaN  2016-10-19T22:05:03Z     False   es  

[5 rows x 22 columns]

In [ ]:
tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14607013 entries, 0 to 14607012
Data columns (total 22 columns):
 #   Column                Dtype  
---  ------                -----  
 0   status_id             int64  
 1   user_id               int64  
 2   created_at            object 
 3   screen_name           object 
 4   text                  object 
 5   source                object 
 6   reply_to_status_id    float64
 7   reply_to_user_id      float64
 8   reply_to_screen_name  object 
 9   is_quote              bool   
 10  is_retweet            bool   
 11  favourites_count      int64  
 12  retweet_count         int64  
 13  country_code          object 
 14  place_full_name       object 
 15  place_type            object 
 16  followers_count       int64  
 17  friends_count         int64  
 18  account_lang          float64
 19  account_created_at    object 
 20  verified              bool   
 21  lang                  object 
dtypes: bool(3), float64(3), int64(6), object

`created_at` column should be in datetime format. `account_lang` is in `float64` format, but intuitively it should be in `object` format.

Column descriptions:
- status_id: unique id for each tweet
- user_id: unique ide for a user
- created_at: datetime for when the tweet was created
- screen_name: user's Twitter handle
- text: content of the tweet
- source: link to the tweet
- reply_to_status_id: tweet id to which this tweet is a reply
- reply_to_user_id: user id to which this tweet is a reply
- is_quote: whether or not this tweet is a quote
- is_retweet: whether or not this tweet is a retweet
- favourites_count: number of users who favorited this tweet
- retweet_count: number of times this tweet was retweeted
- country_code: two-letter code for countries
- place_ful_name: full name of the place if geo-tagged (New York, NY)
- place_type: urban or rural? Can be NaN if not geo-tagged
- followers_count: number of followers this user has
- friends_count: number of people this user follows
- account_lang: the language setting of the user account
- acount_created_at: the date when the account was created
- verified: whether or not the user is verified (i.e., has blue check mark or not)
- lang: the language of the tweet

Let's check what is contained in `account_lang` column.

In [ ]:
tweets['account_lang'].unique()

array([nan])

It only contains NaNs. Given that we are not particularly concerned with what language the user has set their account, this justifies dropping the column.

In [ ]:
tweets.drop(['account_lang'], axis=1, inplace=True)

In [ ]:
# check if account_lang has been dropped
tweets.head()

status_id              user_id            created_at  \
0  1244051646071611394   860252856829587457  2020-03-29T00:00:00Z   
1  1244051645039706112  1125933654943895553  2020-03-29T00:00:00Z   
2  1244051645975191557             80943559  2020-03-29T00:00:00Z   
3  1244051646750928897   817072420947247104  2020-03-29T00:00:00Z   
4  1244051647032102914   788863557349670913  2020-03-29T00:00:00Z   

    screen_name                                               text     source  \
0  IMSS_SanLuis  Ante cualquier enfermedad respiratoria, no te ...  TweetDeck   
1    intrac_ccs  #ATENCIÓN En el Terminal Nuevo Circo se implem...  TweetDeck   
2      rlieving  “People are just storing up. They are staying ...  TweetDeck   
3  Tu_IMSS_Coah  Si empezaste a trabajar, necesitas dar de alta...  TweetDeck   
4  Tabasco_IMSS  Una sociedad informada está mejor preparada an...  TweetDeck   

   reply_to_status_id  reply_to_user_id reply_to_screen_name  is_quote  ...  \
0                 NaN               NaN                  NaN     False  ...   
1                 NaN               NaN                  NaN     False  ...   
2                 NaN               NaN                  NaN     False  ...   
3                 NaN               NaN                  NaN     False  ...   
4                 NaN               NaN                  NaN     False  ...   

   favourites_count  retweet_count  country_code place_full_name place_type  \
0               300              0           NaN             NaN        NaN   
1              1030              1           NaN             NaN        NaN   
2               604              0           NaN             NaN        NaN   
3              1827              0           NaN             NaN        NaN   
4               723              0           NaN             NaN        NaN   

  followers_count  friends_count    account_created_at verified  lang  
0            1008             41  2017-05-04T22:00:38Z    False    es  
1              90            316  2019-05-08T01:21:16Z    False    es  
2             136            457  2009-10-08T21:06:08Z    False    en  
3            1549            170  2017-01-05T18:17:00Z    False    es  
4             868            125  2016-10-19T22:05:03Z    False    es  

[5 rows x 21 columns]

It has been dropped.

Let's check how many missing values are in each column.

In [ ]:
tweets.isna().sum()

status_id                      0
user_id                        0
created_at                     0
screen_name                    2
text                           0
source                        83
reply_to_status_id      12861933
reply_to_user_id        12495830
reply_to_screen_name    12495830
is_quote                       0
is_retweet                     0
favourites_count               0
retweet_count                  0
country_code            13950294
place_full_name         13947685
place_type              13947685
followers_count                0
friends_count                  0
account_lang            14607013
account_created_at             0
verified                       0
lang                           0
dtype: int64

- Two `screen_name` missing
- 83 `source` missing 
- 12,861,933 `reply_to_status_id` missing
- 12,495,830 `reply_to_user_id` missing
- 12,495,830 `reply_to_screen_name` missing
- 13,950,294 `country_code` missing
- 13,947,685 `place_full_name` missing
- 13,947,685 `place_type` missing
- 14,607,013 `account_lang` missing

In [ ]:
tweets.isna().mean()

status_id               0.000000e+00
user_id                 0.000000e+00
created_at              0.000000e+00
screen_name             1.369205e-07
text                    0.000000e+00
source                  5.682202e-06
reply_to_status_id      8.805314e-01
reply_to_user_id        8.554678e-01
reply_to_screen_name    8.554678e-01
is_quote                0.000000e+00
is_retweet              0.000000e+00
favourites_count        0.000000e+00
retweet_count           0.000000e+00
country_code            9.550408e-01
place_full_name         9.548622e-01
place_type              9.548622e-01
followers_count         0.000000e+00
friends_count           0.000000e+00
account_created_at      0.000000e+00
verified                0.000000e+00
lang                    0.000000e+00
dtype: float64

Let's get the unique language codes in the dataset.

In [ ]:
tweets['lang'].unique()

array(['es', 'en', 'ar', 'ur', 'th', 'tl', 'und', 'ca', 'ja', 'fr', 'hu',
       'pt', 'nl', 'it', 'el', 'tr', 'eu', 'ru', 'de', 'sl', 'pl', 'in',
       'hi', 'zh', 'ht', 'ta', 'sv', 'fi', 'kn', 'fa', 'ne', 'si', 'bn',
       'da', 'uk', 'ko', 'ro', 'et', 'no', 'cy', 'vi', 'sd', 'lv', 'ml',
       'ps', 'cs', 'te', 'or', 'gu', 'lt', 'bg', 'ckb', 'mr', 'am', 'lo',
       'is', 'pa', 'sr', 'dv', 'iw', 'my', 'bo', 'hy', 'ka', 'ug', 'km'],
      dtype=object)

As shown, there are many languages in the dataset.

In [ ]:
tweets.loc[tweets['lang']=='en'].count()

status_id               8133785
user_id                 8133785
created_at              8133785
screen_name             8133783
text                    8133785
source                  8133706
reply_to_status_id      1094849
reply_to_user_id        1348976
reply_to_screen_name    1348976
is_quote                8133785
is_retweet              8133785
favourites_count        8133785
retweet_count           8133785
country_code             379717
place_full_name          381809
place_type               381809
followers_count         8133785
friends_count           8133785
account_lang                  0
account_created_at      8133785
verified                8133785
lang                    8133785
dtype: int64

It looks like there are 8,133,785 tweets whose language code is "en" (i.e., English). We want to keep only English tweets for this analysis. There are two reasons for this:
- we are primarily concerned with any presence of bots that may have influenced English-speakers. 
- We may do natural language processing down the line. English language has lots of references to utilize.

Let's only get the rows whose `lang` values are `en`.

In [ ]:
tweets_en = tweets.loc[tweets['lang']=='en']

In [ ]:
tweets_en.to_csv('data/shane_smith/tweets_en.csv')

In [ ]:
tweets_en.shape

(8133785, 21)

`tweets_en` which includes only English tweets contains 8,133785 rows (tweets) and 21 columns.

In [ ]:
tweets_en.head()

status_id     user_id            created_at      screen_name  \
2   1244051645975191557    80943559  2020-03-29T00:00:00Z         rlieving   
6   1244051646935633921    27922157  2020-03-29T00:00:00Z  AmerMedicalAssn   
7   1244051645971025920  1115874631  2020-03-29T00:00:00Z     CGTNOfficial   
9   1244051647149543426   108845320  2020-03-29T00:00:00Z         OnTopMag   
14  1244051645102579712    17673635  2020-03-29T00:00:00Z         livemint   

                                                 text  \
2   “People are just storing up. They are staying ...   
6   .@PatriceHarrisMD spoke with @YahooFinance abo...   
7   First medical team aiding #Wuhan in fight agai...   
9   .@KathyGriffin: @realDonaldTrump Is 'Lying' Ab...   
14  #CoronaUpdate | Johns Hopkins University has s...   

                     source  reply_to_status_id  reply_to_user_id  \
2                 TweetDeck                 NaN               NaN   
6                  Sprinklr                 NaN               NaN   
7      Twitter Media Studio                 NaN               NaN   
9   Twitter for Advertisers                 NaN               NaN   
14                TweetDeck                 NaN               NaN   

   reply_to_screen_name  is_quote  ...  favourites_count  retweet_count  \
2                   NaN     False  ...               604              0   
6                   NaN     False  ...              2894              3   
7                   NaN     False  ...                65             18   
9                   NaN     False  ...              2658              1   
14                  NaN     False  ...               474              2   

    country_code place_full_name place_type followers_count  friends_count  \
2            NaN             NaN        NaN             136            457   
6            NaN             NaN        NaN          714952           6877   
7            NaN             NaN        NaN        14040072             55   
9            NaN             NaN        NaN            5042           5389   
14           NaN             NaN        NaN         1862858            127   

      account_created_at verified  lang  
2   2009-10-08T21:06:08Z    False    en  
6   2009-03-31T17:50:31Z     True    en  
7   2013-01-24T03:18:59Z     True    en  
9   2010-01-27T05:23:15Z    False    en  
14  2008-11-27T09:07:38Z     True    en  

[5 rows x 21 columns]

In [ ]:
tweets_en.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8133785 entries, 2 to 14607012
Data columns (total 21 columns):
 #   Column                Dtype  
---  ------                -----  
 0   status_id             int64  
 1   user_id               int64  
 2   created_at            object 
 3   screen_name           object 
 4   text                  object 
 5   source                object 
 6   reply_to_status_id    float64
 7   reply_to_user_id      float64
 8   reply_to_screen_name  object 
 9   is_quote              bool   
 10  is_retweet            bool   
 11  favourites_count      int64  
 12  retweet_count         int64  
 13  country_code          object 
 14  place_full_name       object 
 15  place_type            object 
 16  followers_count       int64  
 17  friends_count         int64  
 18  account_created_at    object 
 19  verified              bool   
 20  lang                  object 
dtypes: bool(3), float64(2), int64(6), object(10)
memory usage: 1.2+ GB


`created_at` should be in datetime format. `reply_to_source_id` and `reply_to_suer_id` should be in `int` format. `account_created_at` should also be in datetime format. 

Let's check how many values are missing in each column for English tweets.

In [ ]:
tweets_en.isna().sum()

status_id                     0
user_id                       0
created_at                    0
screen_name                   2
text                          0
source                       79
reply_to_status_id      7038936
reply_to_user_id        6784809
reply_to_screen_name    6784809
is_quote                      0
is_retweet                    0
favourites_count              0
retweet_count                 0
country_code            7754068
place_full_name         7751976
place_type              7751976
followers_count               0
friends_count                 0
account_created_at            0
verified                      0
lang                          0
dtype: int64

Convert the counts of missing values into percentages.

In [ ]:
tweets_en.isna().mean()*100

status_id                0.000000
user_id                  0.000000
created_at               0.000000
screen_name              0.000025
text                     0.000000
source                   0.000971
reply_to_status_id      86.539489
reply_to_user_id        83.415151
reply_to_screen_name    83.415151
is_quote                 0.000000
is_retweet               0.000000
favourites_count         0.000000
retweet_count            0.000000
country_code            95.331608
place_full_name         95.305888
place_type              95.305888
followers_count          0.000000
friends_count            0.000000
account_created_at       0.000000
verified                 0.000000
lang                     0.000000
dtype: float64

`country_code`, `place_full_name`, and `place_type` are missing in more than 95% of the rows in the dataset. 86.5% of `reply_to_status_id`, 83.4% `reply_to_user_id`, and 83.4% `reply_to_screen_name` are missing as well. 

Are there any duplicate rows?

In [ ]:
tweets_en.duplicated()

2           False
6           False
7           False
9           False
14          False
            ...  
14607008    False
14607009    False
14607010    False
14607011    False
14607012    False
Length: 8133785, dtype: bool

In [ ]:
# count if there are any True
tweets_en.duplicated().sum()

398

There are 398 duplicated rows. Let's take a look at them.

In [ ]:
# show the rows that are duplicated
duplicate_index = tweets_en.duplicated(keep=False)
tweets_en.loc[duplicate_index, :]

status_id              user_id            created_at  \
154       1244051665642192896           3064266184  2020-03-29T00:00:05Z   
159       1244051665642192896           3064266184  2020-03-29T00:00:05Z   
20425     1244067805265305600           1532718984  2020-03-29T01:04:13Z   
20426     1244067805265305600           1532718984  2020-03-29T01:04:13Z   
55550     1244098805093597184           3326054230  2020-03-29T03:07:24Z   
...                       ...                  ...                   ...   
14527388  1255923136417193992  1141041981390565376  2020-04-30T18:13:04Z   
14550649  1255945376374407168             50023846  2020-04-30T19:41:26Z   
14550650  1255945376374407168             50023846  2020-04-30T19:41:26Z   
14584641  1255980826157879296           1159243226  2020-04-30T22:02:18Z   
14584642  1255980826157879296           1159243226  2020-04-30T22:02:18Z   

              screen_name                                               text  \
154          techjunkiejh  Doors slam shut across borderless Europe as #c...   
159          techjunkiejh  Doors slam shut across borderless Europe as #c...   
20425          marcfavell  @drninashapiro my Condolences from 🇨🇦🇨🇦🇨🇦#Wind...   
20426          marcfavell  @drninashapiro my Condolences from 🇨🇦🇨🇦🇨🇦#Wind...   
55550      mariadestrella  Rhode Island Actively Stopping New Yorkers to ...   
...                   ...                                                ...   
14527388        kate_hosk  @itvnews Best news due to #coronavirus 👏👏👏👏👏\n...   
14550649    JGaryDewberry  #covid19 declared not man-made, so who did it ...   
14550650    JGaryDewberry  #covid19 declared not man-made, so who did it ...   
14584641  Boo_Boo_TheBear  @0164948345aa439 @thomaskaine5 The reason why ...   
14584642  Boo_Boo_TheBear  @0164948345aa439 @thomaskaine5 The reason why ...   

                       source  reply_to_status_id  reply_to_user_id  \
154                    Buffer                 NaN               NaN   
159                    Buffer                 NaN               NaN   
20425     Twitter for Android        1.243960e+18      2.538108e+08   
20426     Twitter for Android        1.243960e+18      2.538108e+08   
55550      Twitter for iPhone                 NaN               NaN   
...                       ...                 ...               ...   
14527388  Twitter for Android        1.255919e+18      2.186694e+07   
14550649      Twitter Web App                 NaN               NaN   
14550650      Twitter Web App                 NaN               NaN   
14584641      Twitter Web App        1.255980e+18      1.159243e+09   
14584642      Twitter Web App        1.255980e+18      1.159243e+09   

         reply_to_screen_name  is_quote  ...  favourites_count  retweet_count  \
154                       NaN     False  ...             14818              0   
159                       NaN     False  ...             14818              0   
20425           drninashapiro     False  ...               763              0   
20426           drninashapiro     False  ...               763              0   
55550                     NaN     False  ...            121499              0   
...                       ...       ...  ...               ...            ...   
14527388              itvnews     False  ...              4521              0   
14550649                  NaN     False  ...             13865              0   
14550650                  NaN     False  ...             13865              0   
14584641      Boo_Boo_TheBear     False  ...               221              0   
14584642      Boo_Boo_TheBear     False  ...               221              0   

          country_code place_full_name place_type followers_count  \
154                NaN             NaN        NaN           28982   
159                NaN             NaN        NaN           28982   
20425              NaN             NaN        NaN             106   
20426              NaN             NaN      

We see that the some of the tweet texts are duplicates, but the tweet IDs (status_id) are still different. Also of note is that the time at which those duplicate tweets were created denoted in `created_at` column are identical, meaning there was no time gap. Also, the `user_id`s are the same for the duplicates. In other words, it was not due to bots copying and pasting (circumventing retweeting) the texts; rather, it was the same user. 

In [ ]:
num_duplicates = 796/2
pct_duplicates = num_duplicates/tweets_en.shape[0]*100
pct_duplicates

0.004893170891534507

Less than 0.005% of all the English tweets are duplicates. I believe they do not hold much significance to sway any outcomes greatly. Additionally, removing them would reduce the burden on the processing unit. 

Maybe not drop the duplicate rows per this source: https://github.com/ttezel/twit/issues/423